chore: federation v2.15 changeset (#3452)
apollographql/federation: docs/source/schema-design/federated-schemas/reference/versions.mdx

- If you maintain a [subgraph-compatible library](/graphos/reference/federation/compatible-subgraphs), consult this article to stay current with recently added directives. All of these directive definitions are also listed in the [subgraph specification](/graphos/reference/federation/subgraph-spec/#subgraph-schema-additions).

## v2.15 (LTS)

<hr/>

| First release | Available in GraphOS? | Minimum router version |
| ------------- | --------------------- | ---------------------- |
| **July 2026** | **Yes** | **`2.16.0`** |

Composition is now written in Rust. No new directives or composition behavior were introduced. Your supergraphs are semantically equivalent to those built with the previous version. The main benefits are faster builds and significantly improved error messages.

Because the Rust implementation is more rigorous, composition now catches several categories of schema problems that were previously inconsistent or missing. If you upgrade and encounter new errors, the following sections explain what to fix.

#### New validations

##### Interfaces implementing `@interfaceObject` now fail explicitly

> Federation doesn't support interfaces as the target of `@interfaceObject`. If your schema uses this pattern, composition now reports `INTERFACE_OBJECT_USAGE_ERROR`.

Author: ```suggestion
Federation doesn't support interfaces implementing `@interfaceObject` interfaces. If your schema uses this pattern, composition now reports `INTERFACE_OBJECT_USAGE_ERROR`.
```

##### Invalid `@override` labels are dropped

If an `@override` directive's `label` references a subgraph name that doesn't exist in your graph, composition now drops that label. Review your `@override` usage to ensure all labels are valid subgraph names.

##### Merged directives on `@external` fields are rejected

> Applying a merged directive to an `@external` field now produces a `MERGED_DIRECTIVE_APPLICATION_ON_EXTERNAL` error. Review your `@external` field definitions..

Author: ```suggestion
Applying a merged directive to an `@external` field now produces a `MERGED_DIRECTIVE_APPLICATION_ON_EXTERNAL` error. Review your `@external` field definitions.
```

Federation v2.13 is a prerequisite for the Connector specification version 0.4.
[Learn more.](/graphos/connectors/reference/changelog)

## v2.12 (LTS)

<hr/>

> 

Reviewer: A little confusing that this adds no new features, and then we go on to list a bunch of differences.  Also, arguably, better validation is a feature.  Maybe this is more clear: 

> This is the first Rust-native Apollo composition release. It introduces no new directives or composition behavior. Rust composition generates semantically equivalent supergraphs to the previous JavaScript binaries. However, the rewrite does surface a number of validation and error-reporting improvements, detailed below, that were previously inconsistent or missing.

Also, we should probably go ahead and review the AI Comments (e.g., headings using sentence case, etc.)
